core/src/services/upyun/lister.rs

```diff
@@ -0,0 +1,100 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+
+use super::core::{ListObjectsResponse, UpYunCore};
+use super::error::parse_error;
+use crate::raw::oio::Entry;
+use crate::raw::*;
+use crate::EntryMode;
+use crate::Metadata;
+use crate::Result;
+
+pub struct UpYunLister {
+    core: Arc<UpYunCore>,
+
+    path: String,
+    limit: Option<usize>,
+}
+
+impl UpYunLister {
+    pub(super) fn new(core: Arc<UpYunCore>, path: &str, limit: Option<usize>) -> Self {
+        UpYunLister {
+            core,
+            path: path.to_string(),
+            limit,
+        }
+    }
+}
+
+#[async_trait]
+impl oio::PageList for UpYunLister {
+    async fn next_page(&self, ctx: &mut oio::PageContext) -> Result<()> {
+        let resp = self
+            .core
+            .list_objects(&self.path, &ctx.token, self.limit)
+            .await?;
+
+        if resp.status() == http::StatusCode::NOT_FOUND {
+            ctx.done = true;
+            return Ok(());
+        }
+
+        match resp.status() {
+            http::StatusCode::OK => {}
+            http::StatusCode::NOT_FOUND => {
+                ctx.done = true;
+                return Ok(());
+            }
+            _ => {
+                return Err(parse_error(resp).await?);
+            }
+        }
+
+        let bs = resp.into_body().bytes().await?;
+
+        let response = serde_json::from_slice::<ListObjectsResponse>(&bs)
+            .map_err(new_json_deserialize_error)?;
+
+        // ref https://github.com/upyun/go-sdk/blob/master/upyun/rest.go#L772
```

Comment:
It's best to provide a link to the API documentation along with some explanation.

Reply:
fixed